### PR TITLE
JAPI-365 Add DFUXRef client

### PR DESCRIPTION
**Pull request by rpastrana**
Signed-off-by: Rodrigo Pastrana <Rodrigo.Pastrana@lexisnexisrisk.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).

## Checklist:
- [x] I have created a corresponding JIRA ticket for this submission
- [x] My code follows the code style of this project.
  - [ ] I have applied the Eclipse code-format template provided.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the HPCC Systems CONTRIBUTORS document (https://github.com/hpcc-systems/HPCC-Platform/wiki/Guide-for-contributors).
- [ ] The change has been fully tested:
  - [ ] I have performed unit tests to cover my changes.
  - [ ] I have performed system test and covered possible regressions and side effects.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.

### CODE REVIEW

**Comment by rpastrana** [comment]
@jpmcmu please review

**Comment by rpastrana** [comment]
@jpmcmu most files in this PR are new generated files, code review should focus on these:

wsclient/src/main/java/org/hpccsystems/ws/client/HPCCWsDFUXRefClient.java
wsclient/src/main/java/org/hpccsystems/ws/client/HPCCWsClient.java
wsclient/pom.xml
wsclient/src/test/java/org/hpccsystems/ws/client/WsDFUXRefClientTest.java
wsclient/src/main/resources/WSDLs/WsDFUXRef-101.wsdl

File: wsclient/src/main/java/org/hpccsystems/ws/client/HPCCWsDFUXRefClient.java [hunk 1]
```diff
@@ -0,0 +1,557 @@
...
+    private static void loadWSDLURL()
+    {
+        try
+        {
+            WSDLURL = getServiceWSDLURL(new WsDFUXRefStub());
+            DEFAULTSERVICEPORT = (new URL(WSDLURL)).getPort();
+        }
+        catch (AxisFault | MalformedURLException e)
+        {
+            log.error("Unable to establish original WSDL URL");
```
Comment (jpmcmu) [inline wsclient/src/main/java/org/hpccsystems/ws/client/HPCCWsDFUXRefClient.java:79]: Do we need to put the client in an "invalid" state here and check for it below, or throw an exception here?

File: wsclient/src/main/java/org/hpccsystems/ws/client/HPCCWsDFUXRefClient.java [hunk 2]
```diff
@@ -0,0 +1,557 @@
...
+            stub = setStubOptions(new WsDFUXRefStub(conn.getBaseUrl() + WSDFUXREFURI), conn);
+        }
+        catch (AxisFault e)
+        {
+            log.error("Could not initialize WsDFUXRefStub - Review all HPCC connection values");
+            e.printStackTrace();
+        }
+        catch (Exception e)
+        {
+            log.error("Could not initialize WsDFUXRefStub - Review all HPCC connection values");
```
Comment (jpmcmu) [inline wsclient/src/main/java/org/hpccsystems/ws/client/HPCCWsDFUXRefClient.java:227]: Same question as above. Do we need to invalidate the client somehow here?

File: wsclient/src/main/java/org/hpccsystems/ws/client/HPCCWsDFUXRefClient.java [hunk 3]
```diff
@@ -0,0 +1,557 @@
...
+        DFUXRefOrphanFilesQueryResponse resp = null;
+
+        try
+        {
+            resp = ((WsDFUXRefStub) stub).dFUXRefOrphanFiles(request);
+        }
+        catch (RemoteException e)
+        {
+            throw new Exception("HPCCWsDFUXRefClient.orphanedFiles(" + cluster + ") encountered RemoteException.", e);
+        }
```
Comment (jpmcmu) [inline wsclient/src/main/java/org/hpccsystems/ws/client/HPCCWsDFUXRefClient.java:294]: Do we need to check the resp for exceptions here as well?

File: wsclient/src/main/java/org/hpccsystems/ws/client/HPCCWsDFUXRefClient.java [hunk 4]
```diff
@@ -0,0 +1,557 @@
...
+
+        try
+        {
+            resp = ((WsDFUXRefStub) stub).dFUXRefLostFiles(request);
+        }
+        catch (RemoteException e)
+        {
+            throw new Exception("HPCCWsDFUXRefClient.lostFiles(" + cluster + ") encountered RemoteException.", e);
+        }
+
```
Comment (jpmcmu) [inline wsclient/src/main/java/org/hpccsystems/ws/client/HPCCWsDFUXRefClient.java:321]: Same question as above, about resp exceptions.

File: wsclient/src/main/java/org/hpccsystems/ws/client/HPCCWsDFUXRefClient.java [hunk 5]
```diff
@@ -0,0 +1,557 @@
...
+        try
+        {
+            resp = ((WsDFUXRefStub) stub).dFUXRefDirectories(request);
+        }
+        catch (RemoteException e)
+        {
+            throw new Exception("HPCCWsDFUXRefClient.directories(" + cluster + ") encountered RemoteException.", e);
+        }
+
```
Comment (jpmcmu) [inline wsclient/src/main/java/org/hpccsystems/ws/client/HPCCWsDFUXRefClient.java:342]: Same question as above, about resp exceptions.

File: wsclient/src/main/java/org/hpccsystems/ws/client/HPCCWsDFUXRefClient.java [hunk 6]
```diff
@@ -0,0 +1,557 @@
...
+        try
+        {
+            resp = ((WsDFUXRefStub) stub).dFUXRefArrayAction(request);
+        }
+        catch (RemoteException e)
+        {
+            throw new Exception("HPCCWsDFUXRefClient.cleanDirectories(" + cluster + ") encountered RemoteException.", e);
+        }
+
```
Comment (jpmcmu) [inline wsclient/src/main/java/org/hpccsystems/ws/client/HPCCWsDFUXRefClient.java:374]: Same question as above, about resp exceptions.

File: wsclient/src/main/java/org/hpccsystems/ws/client/HPCCWsDFUXRefClient.java [hunk 7]
```diff
@@ -0,0 +1,557 @@
...
+        try
+        {
+            resp = ((WsDFUXRefStub) stub).dFUXRefBuildCancel(request);
+        }
+        catch (RemoteException e)
+        {
+            throw new Exception("HPCCWsDFUXRefClient.buildCancel() encountered RemoteException.", e);
+        }
+
```
Comment (jpmcmu) [inline wsclient/src/main/java/org/hpccsystems/ws/client/HPCCWsDFUXRefClient.java:428]: Same question as above, about resp exceptions.

File: wsclient/src/main/java/org/hpccsystems/ws/client/HPCCWsDFUXRefClient.java [hunk 8]
```diff
@@ -0,0 +1,557 @@
...
+        try
+        {
+            resp = ((WsDFUXRefStub) stub).dFUXRefBuild(request);
+        }
+        catch (RemoteException e)
+        {
+            throw new Exception("HPCCWsDFUXRefClient.build(" + cluster +")encountered RemoteException.", e);
+        }
+
```
Comment (jpmcmu) [inline wsclient/src/main/java/org/hpccsystems/ws/client/HPCCWsDFUXRefClient.java:454]: Same question as above, about resp exceptions.

**Review by jpmcmu — APPROVED** [review]
@rpastrana looks good, I don't see any actual issues but do have some questions about error handling.

File: wsclient/src/main/java/org/hpccsystems/ws/client/HPCCWsDFUXRefClient.java (hunk 3 above)
Comment (rpastrana) [reply wsclient/src/main/java/org/hpccsystems/ws/client/HPCCWsDFUXRefClient.java:294]: Good eye- the implementation of the methods is inconsistent, some declare inline exceptions, some don't...

File: wsclient/src/main/java/org/hpccsystems/ws/client/HPCCWsDFUXRefClient.java (hunk 2 above)
Comment (rpastrana) [reply wsclient/src/main/java/org/hpccsystems/ws/client/HPCCWsDFUXRefClient.java]: ensuring the stub is null, and populating the initerr message will invalidate client.

File: wsclient/src/main/java/org/hpccsystems/ws/client/HPCCWsDFUXRefClient.java (hunk 1 above)
Comment (rpastrana) [reply wsclient/src/main/java/org/hpccsystems/ws/client/HPCCWsDFUXRefClient.java:79]: I think this only happens if the URL in the WSDL is invalid, and wouldn't necessarily invalidate running instances of the service, I think we're ok as is.

**Comment by rpastrana** [comment]
@jpmcmu please review

**Comment by jpmcmu** [comment]
@rpastrana looks good

**Pull request merged**